receiver/snmpreceiver/config.go

```diff
@@ -0,0 +1,592 @@
+
+	return combinedErr
+}
+
+// validateEndpoint validates the Endpoint
+func validateEndpoint(cfg *Config) error {
+	if cfg.Endpoint == "" {
+		return errEmptyEndpoint
+	}
+
+	// Ensure valid endpoint
+	u, err := url.Parse(cfg.Endpoint)
+	if err != nil {
+		return fmt.Errorf(errMsgInvalidEndpoint, cfg.Endpoint, err)
+	}
+	if u.Host == "" {
+		return fmt.Errorf(errMsgNoHostInvalidEndpoint, cfg.Endpoint)
+	}
+
+	// Use default port if needed
+	if u.Port() == "" {
+		portSuffix := "161"
+		if cfg.Endpoint[len(cfg.Endpoint)-1:] != ":" {
+			portSuffix = ":" + portSuffix
+		}
+		cfg.Endpoint = cfg.Endpoint + portSuffix
+	}
+
+	// Ensure valide scheme
+	switch strings.ToUpper(u.Scheme) {
+	case "TCP", "TCP4", "TCP6", "UDP", "UDP4", "UDP6": // ok
+	default:
+		return errEndpointBadScheme
+	}
+
+	return nil
+}
+
+// validateVersion validates the Version
+func validateVersion(cfg *Config) error {
+	if cfg.Version == "" {
+		return errEmptyVersion
+	}
+
+	// Ensure valid version
+	switch strings.ToUpper(cfg.Version) {
+	case "V1", "V2C", "V3": // ok
+	default:
+		return errBadVersion
+	}
+
+	return nil
+}
+
+// validateSecurity validates all v3 related security configs
+func validateSecurity(cfg *Config) error {
+	var combinedErr error
+
+	// Ensure valid user
+	if cfg.User == "" {
+		combinedErr = multierr.Append(combinedErr, errEmptyUser)
+	}
+
+	if cfg.SecurityLevel == "" {
+		return multierr.Append(combinedErr, errEmptySecurityLevel)
+	}
+
+	// Ensure valid security level
+	switch strings.ToUpper(cfg.SecurityLevel) {
+	case "NO_AUTH_NO_PRIV":
+		return combinedErr
+	case "AUTH_NO_PRIV":
+		// Ensure valid auth configs
+		return multierr.Append(combinedErr, validateAuth(cfg))
+	case "AUTH_PRIV": // ok
+		// Ensure valid auth and privacy configs
+		combinedErr = multierr.Append(combinedErr, validateAuth(cfg))
+		return multierr.Append(combinedErr, validatePrivacy(cfg))
+	default:
+		return multierr.Append(combinedErr, errBadSecurityLevel)
+	}
+}
+
+// validateAuth validates the AuthType and AuthPassword
+func validateAuth(cfg *Config) error {
+	var combinedErr error
+
+	// Ensure valid auth password
+	if cfg.AuthPassword == "" {
+		combinedErr = multierr.Append(combinedErr, errEmptyAuthPassword)
+	}
+
+	// Ensure valid auth type
+	if cfg.AuthType == "" {
+		combinedErr = multierr.Append(combinedErr, errEmptyAuthType)
+	} else {
+		switch strings.ToUpper(cfg.AuthType) {
+		case "MD5", "SHA", "SHA224", "SHA256", "SHA384", "SHA512": // ok
+		default:
+			combinedErr = multierr.Append(combinedErr, errBadAuthType)
+		}
+	}
+
+	return combinedErr
+}
+
+// validatePrivacy validates the PrivacyType and PrivacyPassword
+func validatePrivacy(cfg *Config) error {
+	var combinedErr error
+
+	// Ensure valid privacy password
+	if cfg.PrivacyPassword == "" {
+		combinedErr = multierr.Append(combinedErr, errEmptyPrivacyPassword)
+	}
+
+	// Ensure valid privacy type
+	if cfg.PrivacyType == "" {
+		combinedErr = multierr.Append(combinedErr, errEmptyPrivacyType)
```

Comment:
Same here this could just return so the switch case doesn't have to be in an `else`.

Reply:
changed